### PR TITLE
(#825) Binary

### PR DESCRIPTION
**Pull request by borysfan**
This PR is for #825 issue. It covers:
- new Binary class
- unit tests for Binary class

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#1054](https://github.com/yegor256/cactoos/issues/1054) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/93d7f727-cfb4-4d9d-97ab-7df92efc4871); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#1054` already assigned to @fabriciofx, can't assign to @fabriciofx

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.scalar;
+
+import org.cactoos.Scalar;
+
+/**
+ * Binary  operation.
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan `Binary operation.` (there is more than one whispace here).

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * <pre>{@code
+ * final AtomicInteger counter = new AtomicInteger();
+ * new Binary<>(
+ *     () -> true,
+ *     () -> counter.incrementAndGet
+ * ).value() // will be equal true
+ * }</pre>
+ *
+ * @since 0.9
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Use `1.0` instead.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * <p>This class implements {@link Scalar}, which throws a checked
+ * {@link Exception}. This may not be convenient in many cases. To make
+ * it more convenient and get rid of the checked exception you can
+ * use the {@link UncheckedScalar} decorator. Or you may use
+ * {@link IoCheckedScalar} to wrap it in an IOException.</p>
+ *
+ * <pre>{@code
+ * final AtomicInteger counter = new AtomicInteger();
+ * new Binary<>(
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Will `Binary` use generics?

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ *     () -> true,
+ *     () -> counter.incrementAndGet
+ * ).value() // will be equal true
+ * }</pre>
+ *
+ * @since 0.9
+ */
+public final class Binary implements Scalar<Boolean> {
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Please, use `Binary<T>`.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Scalar for condition.
+     */
+    private final Scalar<Boolean> condition;
+    /**
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Please, put one more new line here.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+     * Scalar executed when condition is true.
+     */
+    private final Scalar<?> consequent;
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Please, change it to `Scalar<T>`.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Ctor.
+     *
+     * @param condition Boolean function
+     * @param runnable Runnable executed when condition is true
+     */
+    public Binary(final Scalar<Boolean> condition, final Runnable runnable) {
+        this(condition, () -> {
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Please, break the line here:
```
this(
    condition...
```

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            runnable.run();
+            return null;
+        });
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param condition Boolean function
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan `conditition` isn't a function. Please, change it to a properly description.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+     * @param consequent Scalar executed when condition is true
+     */
+    public Binary(final Scalar<Boolean> condition, final Scalar<?> consequent) {
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Please, change to `Scalar<T>`.

File: src/test/java/org/cactoos/scalar/BinaryTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        new Assertion<>(
+            "Binary must not to invoke increment method",
+            counter::get,
+            new IsEqual<>(expected)
+        ).affirm();
+    }
+
+    @Test
+    public void throwsException() {
+        final String msg = "custom exception";
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/BinaryTest.java]: @borysfan Please, can you improve this exception message? And start it with a capital letter.

**Comment by fabriciofx** [comment]
@borysfan ping,

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 3 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: The plan was to not use generics. I missed it in javadoc description.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 4 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: Does it make sense to use generics here? Binary always returns Boolean.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 6 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: The same question as before. Does it make sense to use generics here? Binary always returns Boolean.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 2 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: Done.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 1 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: Done.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 5 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: Done.

File: src/test/java/org/cactoos/scalar/BinaryTest.java (hunk 10 above)
Comment (borysfan) [reply src/test/java/org/cactoos/scalar/BinaryTest.java]: Done.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 8 above)
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: Done.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 3 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/scalar/Binary.java]: @borysfan I'm sorry, but I was thinking in `Binary` as a sort of `Ternary`. You're completely right.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 4 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Same above.

File: src/main/java/org/cactoos/scalar/Binary.java (hunk 6 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/scalar/Binary.java]: @borysfan Same above.

File: src/main/java/org/cactoos/scalar/Binary.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+     * Ctor.
+     *
+     * @param condition Boolean The condition
+     * @param runnable Runnable executed when condition is true
+     */
+    public Binary(final Scalar<Boolean> condition, final Runnable runnable) {
+        this(
+            condition, () -> {
+                runnable.run();
+                return null;
```
[the remark this answers is not in the document]
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: @fabriciofx I do not like it too. I will try to figure something out.

[the remark this answers is not in the document]
Comment (borysfan) [reply src/main/java/org/cactoos/scalar/Binary.java]: @fabriciofx I changed null into 0. I am not sure how can we make it better.

**Comment by borysfan** [comment]
@fabriciofx I started to think if passing Scalar to a Binary makes sense. Do you see any usage of such api? I am afraid that we want to solve somerhing which is wrong on the task assumption. What do you think about it?

**Comment by llorllale** [comment]
@borysfan make `Binary` accept a `Proc` instead of a `Scalar` for the "action"

**Comment by llorllale** [comment]
@borysfan ping

**Comment by borysfan** [comment]
@llorllale @fabriciofx I replaced Scalar with Proc. Could please look at it once again?

**Comment by borysfan** [comment]
@llorllale @fabriciofx ping

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
Code review was too long (20 days), architects (@llorllale) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
The job [#1054](https://github.com/yegor256/cactoos/issues/1054) is now out of scope

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Pull request merged**